src/main/java/de/hhu/stups/plues/ui/controller/Reports.java

```diff
@@ -185,67 +270,242 @@ private void displayModuleAbstractUnitUnitSemesterConflicts(final Store store,
+  private Map<Module, Set<AbstractUnit>>
+      displayQuasiMandatoryModuleAbstractUnits(final Store store, final ReportData reportData) {
+    Map<Module, Set<AbstractUnit>> quasiMandatoryModuleAbstractUnits =
+        reportData.getQuasiMandatoryModuleAbstractUnits()
         .entrySet().stream().collect(Collectors.toMap(
           entry -> store.getModuleById(entry.getKey()),
           entry -> entry.getValue().stream().map(
-              store::getAbstractUnitById).collect(Collectors.toSet()))));
+          store::getAbstractUnitById).collect(Collectors.toSet())));
+    this.quasiMandatoryModuleAbstractUnits.setData(quasiMandatoryModuleAbstractUnits);
+
+    return quasiMandatoryModuleAbstractUnits;
   }
 
-  private void displayMandatoryModules(final Store store, final ReportData reportData) {
-    mandatoryModules.setData(reportData.getMandatoryModules()
+  private Map<Course, Set<Module>> displayMandatoryModules(final Store store,
+                                                           final ReportData reportData) {
+    Map<Course, Set<Module>> mandatoryModules = reportData.getMandatoryModules()
         .entrySet().stream().collect(Collectors.toMap(
           entry -> store.getCourseByKey(entry.getKey()),
           entry -> entry.getValue().stream().map(
-              store::getModuleById).collect(Collectors.toSet()))));
+          store::getModuleById).collect(Collectors.toSet())));
+    this.mandatoryModules.setData(mandatoryModules);
+
+    return mandatoryModules;
   }
 
-  private void displayImpossibleCourses(final Store store, final ReportData reportData) {
-    impossibleCourses.setData(reportData.getImpossibleCourses()
-          .stream().map(store::getCourseByKey).collect(Collectors.toList()),
+  private List<List<Course>> displayImpossibleCourses(final Store store,
+                                                                 final ReportData reportData) {
+    List<Course> impossibleCourses = reportData.getImpossibleCourses()
+        .stream().map(store::getCourseByKey).collect(Collectors.toList());
+    List<Course> impossibleCoursesBecauseOfImpossibleModules =
         reportData.getImpossibleCoursesBecauseofImpossibleModules()
-          .stream().map(store::getCourseByKey).collect(Collectors.toList()),
+          .stream().map(store::getCourseByKey).collect(Collectors.toList());
+    List<Course> impossibleCoursesBecauseOfImpossibleModuleCombinations =
         reportData.getImpossibleCoursesBecauseOfImpossibleModuleCombinations()
-          .stream().map(store::getCourseByKey).collect(Collectors.toList()));
+          .stream().map(store::getCourseByKey).collect(Collectors.toList());
+    this.impossibleCourses.setData(impossibleCourses,
+        impossibleCoursesBecauseOfImpossibleModules,
+        impossibleCoursesBecauseOfImpossibleModuleCombinations);
+
+    return Arrays.asList(impossibleCourses, impossibleCoursesBecauseOfImpossibleModules,
```

Comment:
split into a method that computes the data (e.g. in RealReportData) and one step that displays it.